docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+
+The following data types shall be supported by the IPC mechanism:
+
+* **Primitive Types**:
+
+  * Boolean
+  * Numeric (fixed-size integers 8-128 bits, signed and unsigned; IEEE 754 floating-point numbers)
+
+* **Sequence Types**:
+
+  * Array (fixed-length)
+
+* **User-Defined Types**:
+
+  * Struct
+  * Tuple
+  * Enum (tag-only)
+  * Variant ("tagged union"; *optional*)
+
+* **Fixed-Size, Variable-Length Containers**:
+
+  * Vector
+  * Queue
+  * Hash map (*optional*)
+  * Hash set (*optional*)
+  * Binary tree (*optional*)
+
+* **Specialized Variants**:
+
+  * Result
+  * Option
+
+All provided data types must ensure fixed size and consistent memory layouts.
+
+Type Description
+----------------
+
+A critical scalability feature involves gateway processes, which subscribe to IPC endpoints and translate ABI-compatible data types into external serialization formats. These gateways require the ability to interpret data without compile-time access to type definitions. To address this, an explicit runtime-readable type description format is necessary. This description allows dynamic, runtime interpretation of data structures, enabling the addition of new IPC topics without recompiling gateway processes.
+
+In summary, the motivation behind this feature request is to define and standardize ABI-compatible data types and a runtime-accessible type description mechanism to ensure interoperability and scalability in zero-copy IPC scenarios involving multiple languages and dynamic environments.
+
+
+.. Rationale
+.. ==========
+
+
+Specification
+=============
+
+ABI Compatibility
+-----------------
+
+This specification defines the set of rules and constraints for representing data types in shared memory such that they can be interpreted consistently across processes implemented in C++17 and Rust 1.8x. These types enable zero-copy inter-process communication by enforcing ABI compatibility at the memory layout level. The focus is on data exchange between processes using the same endianness.
+
+Assumptions
+^^^^^^^^^^^
+
+* Shared memory regions are mapped at correctly aligned virtual addresses in both processes.
+* No serialization or runtime copying occurs when interpreting a type from memory.
+* Processes use the same endianness.
+* No synchronization or atomicity guarantees are defined at the data type level; these are provided by the IPC framework.
+* All memory is allocated statically or pre-reserved. Dynamic memory allocation is disallowed.
+
+Type Conformance
+^^^^^^^^^^^^^^^^
+
+Types used in shared memory must meet the following criteria:
+
+1. **Fixed size and alignment**: Every type must have a known, constant size and alignment at compile time.
+2. **Consistent layout across languages**: The layout of a type must be identical in Rust and C++, and on all platforms.
+3. **No absolute pointers or references**: Types must not contain pointers to heap memory, function pointers, or references.
```

Comment:
A pointer to stack is equally bad.
Also wording here is tricky. What we really want to say is that pointers and even references are fine, as long as they are implemented with a relative offset and do not point outside of the transferred data.


```suggestion
3. **No absolute pointers or references**: Types must not contain absolute pointers/function pointers/references that point out of the transferred data.
```

Reply:
Sorry my "improved wording" is also wrong again...

```suggestion
3. **No absolute pointers or references**: Types must not contain absolute pointers/function pointers/references or any pointer/reference that point out of the transferred data.
```